### PR TITLE
Second part of refactoring the routing part

### PR DESCRIPTION
**Pull request by ChristinaZ**
…PdlOverlapWithNext;Remove DeepSeekV3 float32 logits constraint from kernel launchers

## 📌 Description

1. Add dynamic block kernel (`routingIndicesDynBlockKernel`) comes from the TensorRT-LLM. https://github.com/NVIDIA/TensorRT-LLM/pull/12456/ . Made related modification by refactoring  `LAUNCH_ROUTING_CUSTOM` with `dispatchRoutingPolicy` and `queryDispatchedMaxExperts`
2. Simplify PDL (Programmatic Dependent Launch) for routing kernels, as the bug related to PDL is solved.
3. Added a default fallback tier (`Tier<1024, 32>`) to support future models with >512 experts using the DeepSeek nGroup≤1 / MiniMax2 routing policy.
4. Remove DeepSeekV3 float32 logits constraint
5. Improve policy tier dispatch error messages

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests
```
python -m pytest tests/moe/test_trtllm_gen_fused_moe.py -k "test_dyn_block_kernel_routing or test_tier_1024_experts_routing" -xvs
python3 -m pytest tests/moe/test_trtllm_gen_fused_moe.py -k "test_routing_dtype_flexibility" -xvs
```

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Dynamic single-block routing for small token/expert workloads to improve performance.

* **Improvements**
  * Added a 1024-expert policy tier and clearer tier-dispatch diagnostics.
  * More flexible routing/logits dtype handling and simplified kernel overlap/launch synchronization.
  * Automatic histogram-thread sizing and removal of legacy cluster-size/overlap public flags.
  * Autotuner now initializes packed TopK with per-token unique expert IDs.

* **Tests**
  * Added tests for dynamic-block routing, 1024-expert tier, MiniMax2 routing, and routing dtype combinations.

### CODE REVIEW

File: tests/moe/test_trtllm_gen_fused_moe.py
Comment (jiahanc) [inline tests/moe/test_trtllm_gen_fused_moe.py:1]: nit: Would moving the routing test to a new test file make sense?

Comment (ChristinaZ) [reply tests/moe/test_trtllm_gen_fused_moe.py:1]: We should probably have a dedicated unit test for routing. So far we’ve only tested it together with GEMM and the final part. I’ll figure out a separate test in a follow-up PR.

**Review by jiahanc — APPROVED** [review]
LGTM. Thanks for contribution!

**Comment by jiahanc** [comment]
@flashinfer-bot run

**Comment by nvpohanh** [comment]
> Add dynamic block kernel (routingIndicesDynBlockKernel) comes from the TensorRT-LLM. https://github.com/NVIDIA/TensorRT-LLM/pull/12456 . Made related modification by refactoring  LAUNCH_ROUTING_CUSTOM with dispatchRoutingPolicy and queryDispatchedMaxExperts
> Simplify PDL (Programmatic Dependent Launch) for routing kernels, as the bug related to PDL is solved.
> Added a default fallback tier (Tier<1024, 32>) to support future models with >512 experts using the DeepSeek nGroup≤1 / MiniMax2 routing policy.
> Remove DeepSeekV3 float32 logits constraint
> Improve policy tier dispatch error messages

cc @leejnau @trevor-m for vis since this is related to DSV3 and MiniMax2.

File: tests/moe/test_trtllm_gen_fused_moe.py
Comment (ChristinaZ) [reply tests/moe/test_trtllm_gen_fused_moe.py:1]: We should probably have a dedicated unit test for routing. So far we’ve only tested it together with GEMM and the finalize part. I’ll figure out a separate test in a follow-up PR.

File: flashinfer/fused_moe/core.py [hunk 1]
```diff
@@ -1264,6 +1197,34 @@ def forward(
...
+                    DynamicTensorSpec(
+                        (0, 1, 2, 3, 4),
+                        (0, 0, 0, 0, 0),
+                        get_last_power_of_2_num_tokens_buckets(tune_max_num_tokens, 1),
+                        lambda x: min(last_positive_power_of_2(x), tune_max_num_tokens),
+                        cls.dynamic_tensor_initializers[:5],
+                    ),
+                ),
+                **kwargs,
+            )
```
[the remark this answers is not in the document]
Comment (ChristinaZ) [reply flashinfer/fused_moe/core.py]: It's a bug I introduced during rebase. Fixed.

**Review by aleozlx — APPROVED** [review]
